modules/CbChargeSOMDriver/chargesom/CbChargeSOM.cpp

```diff
@@ -0,0 +1,253 @@
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+            return types::cb_board_support::CPState::B;
+        default:
+            return types::cb_board_support::CPState::C;
+        }
+    }
+
+    if (voltage >= 5000 /* mV */) /* 5 V <= x < 7 V */
+        return types::cb_board_support::CPState::C;
+
+    if (voltage >= 4000 /* mV */) { /* 4 V <= x < 5 V */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+            return types::cb_board_support::CPState::C;
+        default:
+            return types::cb_board_support::CPState::D;
+        }
+    }
+
+    if (voltage >= 2000 /* mV */) /* 2 V <= x < 4 V */
+        return types::cb_board_support::CPState::D;
+
+    if (voltage >= 1000 /* mV */) { /* 1 V <= x < 2 V */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+        case types::cb_board_support::CPState::D:
+            return types::cb_board_support::CPState::D;
+        default:
+            return types::cb_board_support::CPState::E;
+        }
+    }
+
+    if (voltage >= -1000 /* mV */) /* -1 V <= x < 1 V */
+        return types::cb_board_support::CPState::E;
+
+    if (voltage >= -2000 /* mV */) { /* -2 V <= x < -1 V (not standard) */
+        switch (previous_state) {
+        case types::cb_board_support::CPState::A:
+        case types::cb_board_support::CPState::B:
+        case types::cb_board_support::CPState::C:
+        case types::cb_board_support::CPState::D:
+        case types::cb_board_support::CPState::E:
+            return types::cb_board_support::CPState::E;
+        default:
+            return types::cb_board_support::CPState::PilotFault;
+        }
+    }
+
+    if (voltage >= -10000 /* mV */) /* -10 V <= x < -2 V */
+        return types::cb_board_support::CPState::PilotFault;
+
+    if (voltage >= -11000 /* mV */) { /* -11 V <= x < -10 V (not standard) */
+        return previous_state == types::cb_board_support::CPState::F ? types::cb_board_support::CPState::F
+                                                                     : types::cb_board_support::CPState::PilotFault;
+    }
+
+    if (voltage >= -13000 /* mV */) /* -13 V <= x < -11 V */
+        return types::cb_board_support::CPState::F;
+
+    return types::cb_board_support::CPState::PilotFault; /* < -13 V */
+}
+
+double CbChargeSOM::cp_get_duty_cycle() const {
+    // this method must be called with `ctx_mutex` already held
+
+    return this->ctx.data.duty_cycle;
+}
+
+void CbChargeSOM::cp_set_duty_cycle(double duty_cycle) {
+    std::scoped_lock lock(this->ctx_mutex);
+
+    this->ctx.data.duty_cycle = floorl(duty_cycle);
```

Comment:
The safety controller only supports integer PWM? That's okay for DC, but quite unfortunate for (future) AC basic charging.

Reply:
Yes, for the moment. Please keep in mind, that this is still far away from anything we will use later. This is only one of the open points we have to coordinate later with the safety fw team.